frb_example/pure_dart/dart/lib/bridge_generated.dart

```diff
@@ -2124,6 +2610,63 @@ ConcatenateWith _wire2api_concatenate_with(FlutterRustBridgeExampleSingleBlockTe
   );
 }
 
+CustomError _wire2api_custom_error(dynamic raw) {
+  final access = raw is Map ? raw['dart_object'] : raw;
+  switch (access[0]) {
+    case 0:
+      return Error0(_wire2api_String(access[1]), raw['backtrace']);
```

Comment:
Currently, if I understand correctly, we are passing backtrace from rust to dart using an extra mechanism (this raw['backtrace']). 

However, we already have excellent tooling to pass arbitrary data (struct etc) from rust to dart - the frb itself!

So, maybe we do not need to manually treat backtrace as something very special, i.e. simply the implementation.

For example, we may create a `struct CustomErrorWithBacktrace{inner:CustomError, backtrace:Backtrace}`, and then let the existing frb tooling to automatically generate everything for it. Then, we only need to simply "box" a CustomError into a CustomErrorWithBacktrace etc - which is much type-safer and easier to write and less error-prone I guess.

Reply:
Just tried this, but remember:

```dart
@freezed
class CustomError with _$CustomError implements FrbException {
  @Implements<FrbBacktracedException>()
  const factory CustomError.error0(
    String field0,
  ) = Error0;
  @Implements<FrbBacktracedException>()
  const factory CustomError.error1(
    int field0,
  ) = Error1;
}
```

this does not work, we need a `Backtrace backtrace` so in the other end, `Error0` and `Error1` will implement `FrbBacktracedException`. That's why I pass a separate backtrace. What do you think?

Reply:
Then what about this:

```
enum CustomErrorWrapped {
  Error0 { sth: String, backtrace: Backtrace, },
...
}
```

then let frb do all the heavy lifting of code generation?

Reply:
Where would we put a backtrace in

```
pub enum CustomError {
    Error0(String),
    Error1(u32),
}
```

?

Reply:
Ops, didn't see your comment.

So,

```
enum CustomErrorWrapped {
  Error0 { sth: String, backtrace: Backtrace, },
...
}
```

is ok I guess

Reply:
Hmmm, I think it's strange to manually throw an error like this

```rust
pub enum CustomError {
    Error0{e: String, backtrace: Backtrace},
    Error1{e: u32, backtrace: Backtrace},
}

pub fn return_err_custom_error() -> Result<u32, CustomError> {
    Err(Error0{e: CustomError::Error1(3), backtrace: Backtrace})
}
```

maybe we should leave the backtrace for the user to not care about

Reply:
Ops, I meant
```rust
pub enum CustomError {
    Error0{e: String, backtrace: Backtrace},
    Error1{e: u32, backtrace: Backtrace},
}

pub fn return_err_custom_error() -> Result<u32, CustomError> {
    Err(CustomError::Error0{e: "".into(), backtrace: ?})
}
```

It's much better to always do `CustomError::Error0("".into())`. Very little people are going to care about backtraces, but if we pass them automatically, they will be happy to see the backtrace once something crashes. When I used this lib I was very confused about the "Result unwrap" panic without any line errors. I think the backtrace should always appear

Reply:
I very agree with the backtrace - the first thing I do is to add backtrace support in my own app :)

Well, I am originally thinking about this: For an end user, just write down

```rust
use thiserror::Error;

#[derive(Error)]
pub enum CustomError {
    Error0{some_field: String},
    Error1{maybe_another_field_name: u32},
}

fn f() { Err(CustomError::Error0("oops") }
```

and in dart it automatically provides backtrace.

However, after looking at https://docs.rs/thiserror/1.0.31/thiserror/, I realized, weirdly, thiserror does not provide some backtrace magic at all...?

Then, seems that we have to let the user manually capture the backtrace: https://docs.rs/backtrace/latest/backtrace/struct.Backtrace.html#method.new (this crate for stable; for nightly we can use std::backtrace but we should not target nightly)

So yes, seems that we have to 

```rust
pub enum CustomError {
    Error0{e: String, backtrace: Backtrace},
    Error1{e: u32, backtrace: Backtrace},
}

pub fn return_err_custom_error() -> Result<u32, CustomError> {
    Err(CustomError::Error0{e: "".into(), backtrace: backtrace::Backtrace::new()})
}
```

Reply:
Given that `thiserror` and `anyhow` seems to be de-facto, this approach may be acceptable by Rust users...

Reply:
Indeed, if we use this approach, my proposal above about the "automatically created backtrace field" just no longer holds, so maybe ignore it.

Btw, backtrace should be captured (i.e. `Backtrace::new()`) at the *place where original error occurs* (i.e. when user really causes an exception), instead of at the place where we transfer errors (e.g. frb handler.rs) - indeed there is no way to know the backtrace if it is not captured when the error occurs.

Reply:
Just tried and `std::backtrace::Backtrace` is too complex, it has an UnsafeCell, so frb isn't able to parse it into a dart object. Any suggestions? 

`thread 'main' panicked at 'parse_type failed for ty=Backtrace', frb_codegen/src/parser/ty.rs:146:32`

Reply:
> Just tried and std::backtrace::Backtrace is too complex, it has an UnsafeCell, so frb isn't able to parse it into a dart object

Surely not parse into *object*, but we should let it format itself into a string, or `Vec<String>` or sth like that. Indeed humans only read strings :)

Reply:
I don't get it. thiserror and anyhow surely support the std::backtrace::Backtrace, but not other backtrace formats I guess. Shouldn't I put Backtrace in the error?

Reply:
Also, we would have to add `backtrace: Backtrace` to all variants. I find this strange

Reply:
> I don't get it. thiserror and anyhow surely support the std::backtrace::Backtrace, but not other backtrace formats I guess. Shouldn't I put Backtrace in the error?

Well I mean, in rust side, it is Backtrace; in dart side, it is a string.

> Also, we would have to add backtrace: Backtrace to all variants. I find this strange

Seems this is how we *have to* do it if people are using thiserror?

Reply:
> Well I mean, in rust side, it is Backtrace; in dart side, it is a string.

How?

> Seems this is how we have to do it if people are using thiserror? So it is thiserror that is strange... (Btw I also feel strange)

I think this is going to not be used, even by me. What do you think of catching the backtrace with catch_unwind at the moment of the taskfn call?

Reply:
> How?

What about implementing custom logic when "converting Backtrace from Rust-api to Rust-wire types"?

>  What do you think of catching the backtrace with catch_unwind at the moment of the taskfn call?

Well that is not the same thing.